source/common/http/conn_manager_impl.h

```diff
@@ -203,13 +227,25 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
       parent_.decodeData(this, *parent_.buffered_request_data_, end_stream,
                          ActiveStream::FilterIterationStartState::CanStartFromCurrent);
     }
+    void doMetadata() override { drainSavedRequestMetadata(); }
+    MetadataMapVector& saved_metadata() override { return *getSavedRequestMetadata(); }
     void doTrailers() override { parent_.decodeTrailers(this, *parent_.request_trailers_); }
     const HeaderMapPtr& trailers() override { return parent_.request_trailers_; }
 
+    void drainSavedRequestMetadata() {
+      for (auto& metadata_map : *getSavedRequestMetadata()) {
```

Comment:
Perhaps ASSERT that the storage vector pointer is not nullptr so we don't end up creating an empty vector anyway? Can you audit other calls to make sure we aren't accidentally creating a vector just to check that it's empty?

Reply:
Good point! Fixed in both request and response direction.

Reply:
Should we have a parallel assert in drainSavedResponseMetadata?

Reply:
Fixed. Thanks!